### ISSUE TITLE
Potential issue introduced recently

### ISSUE THREAD

**Issue by silviucpp** [mysql-otp/mysql-otp#95]
Hello,

I upgraded the entire cluster we have to OTP 21 version of mysql-otp and we observed that in case the cluster have a problem (network issues) the driver stays forever in start_transaction or rollbacks.

I see some changes related to this here:

https://github.com/mysql-otp/mysql-otp/pull/91/files

And I'm wonder why you used 'infinity' ? I mean you could increase the timeout but infinity seems a bit risky.

**Comment by silviucpp** [comment]
Just to have an idea of our situation: We have a percona cluster with master-master replication (based on mysql 5.6) in a datacenter where always had issues with connectivity between machines. Let's make an abstraction to the fact that the DC should't have any issue :)) for the money we pay we cannot comment on this.

So far we used for more than 1 year mysql-otp and we didn't have any issue, those network problems were very well handled (including our app that retries the failed queries). Last version we used and wroked great was ef48ab3d545049acdaa73f2e17f0abd0a4a98b6d . 4 weeks ago we updated to master (was the only change required in the code to migrate to OTP 21), also we updated OS packages and OTP version. 

Since this update we had 2 major issues where:
- we observers driver get stucked in start_transaction or rollbacks
- we don't get any error for simple queries (INSERTS that returns ok) but the database rollbacks the transactions (we cannot find any longer the data on the server when we query). When this issues occurs I see deadlocks and transaction-wait timeouts a lot.
  
In this moment  have no clear way on how to start the investigation, Usually in a percona master-master replication it's almost impossible to loose data and because we didn't done this in last 2-3 years since we are using same db I will start by investigating the driver changes.

For this reason I reverted back to ef48ab3d545049acdaa73f2e17f0abd0a4a98b6d and only applied the OTP 21 patch (changes related to the way you get the erlang stacktrace). I will try to observer if we still have problems when such networking issues happens. If not I will try to put the other patches one by one and find which one breaks. Do you have any other idea ?

Silviu

**Comment by zuiderkwast** [comment]
> Ok I see but is not a better idea just to kill the connection in case a timeout occurs ? In this case the pool library will create a new connection and also the server will cleanup as well the pending transactions ?

@emfa Do you have an opinion on this?

**Comment by silviucpp** [comment]
What do you think about:

```erl
  case catch mysql:transaction(Conn, Fun, Args, Retries) of
        {atomic, _} = S ->
            S;
        {aborted, _} = A ->
            A;
        Error ->
            ?ERROR_MSG("mysql:transaction unexpected response for connection: ~p error: ~p", [Conn, Error]),
            exit(Conn, kill),
            Error
    end.
```

Looking to the code I think it's much safe to assume that in case you didn't received `{atomic, _}` or `{aborted, _}` it's better just to kill the connection and the pool will recreate it.

@zuiderkwast what do you think my assumption is correct ?

**Comment by zuiderkwast** [comment]
Yes, it should be safe.

But `transaction` can never return anything else. There can be an exception from `transaction` if there is an exception in `Fun`, but in this case the transaction is rollbacked (using catch and re-raise) before the exception reaches your `catch` so I don't see how this can solve your problem. But you can try it, sure.

**Comment by zuiderkwast** [comment]
This is a potential cause of your problem: The application process dies when the connection is in a transaction.

When you (application process) checks out a connection from Poolboy, Poolboy monitors you. If you die while you have a connection checked out, Poolboy takes the connection back and puts it back in the pool. Poolboy doen't know that the connection can be in a bad state.

```  
              Poolboy
               /   \
              /     \
      monitor/       \link
            /         \
           /           \
          /             \
         v               \
Application              MySQL
  process  <----------> connection
             queries
```

Do you think this is what happens?

@emfa suggested in #90 that the connection (mysql gen_server) should add a monitor to the application process when a transaction is started, so that the MySQL process can kill itself (or at least abort the transaction) before poolboy takes it back into the pool. What do you think about that?

**Comment by silviucpp** [comment]
So 100 % sure my issue comes from the fact that connection is in bad state. I mean a transaction started but commit never had the chance to execute and all inserts are returning ok and once connection dies everything gets reverted. 

Basically all my mysql calls comes from a cowboy handler process and there I have a timeout of 10 seconds. In case the request is not executed in this time the cowboy process (application process) in your case dies. and you are right: poolboy adds it back in the pool. And that connection is in a bad state.

The fact that in the last version of mysql pool an infinity timeout was added on start transaction, commit and rollback make things even worst for this case (where application process dies in a certain timeframe).

If I revert that change with infinity then transaction can also return the timeouts and the code I pasted above catches and kills the connection (I tested by adding a time:sleep(10000) inside the start_transaction call). And seems this timeouts generated by start_transaction, commit and rollback are catched properly and connection is restarted. 

Also in my case because infinity was removed, the transaction call returns early than the cowboy kills the app process, but still that problems remains and what emfa proposed seems good idea.

Do you got my point ?

**Comment by silviucpp** [comment]
So to be clear, reverting the commit with inifinity for gen_server:call inside transaction, makes that function to throw  ` {'EXIT',{timeout,{gen_server,call,[<0.124.0>,start_transaction]}}}` which is catched by the method I pasted above and connection gets killed, while timeout is respected as well.

**Comment by zuiderkwast** [comment]
Yes, I get it. So, let's add a monitor and save it in the gen_server state! Demonitor after commit/rollback. Maybe I can get time to do it in the weekend, but I'd be happy if you or someone else wants to do it. I can only work on this projecton my spare time unfortunately.

**Comment by silviucpp** [comment]
@zuiderkwast please take a look here: https://github.com/mysql-otp/mysql-otp/pull/96
What do you think ? Please provide some feedback and after this I'll add a test for it as well in case you think it's fine.

**Comment by emfa** [comment]
I'm currently refactoring alot of code so that one process owns the connection and is thus linked to it. Because that seems to be the way to avoid all issues, and then I still think infinitiy is correct. If any of them fail, the best thing is that both die.

If one were to improve the driver I think that a stack of monitors for each nested transaction call might be the way to do it properly.

**Comment by emfa** [comment]
The fixes introduces have mitigated almost all of our errors though so I still stand by them as it is currently.

**Comment by emfa** [comment]
@silviucpp Link to the connection, if the cowboy process dies, the connection dies and poolboy will create a new one? That might not work for high churn loads have I've experienced though.

**Comment by zuiderkwast** [comment]
@emfa OK, your scenario should work as well. We have to acknowledge that there are several usage scenarios. But the monitor solution doesn't hurt, does it? If the owner (the process calling start_link) dies, the connection dies too even before it receives DOWN from the monitor. That's fine I guess. Don't you think?

**Comment by zuiderkwast** [comment]
Considering this solved by #96.

**Comment by zuiderkwast** [comment]
With the monitors introduced, I don't see how a connections can possibility end up back in the pool in a bad state now, with or without infinity. If there is any other timeout, such as your cowboy worker timeout of 10 seconds, the transaction will be aborted by the monitor when the cowboy worker process dies.

I also don't see any possibility that transaction will return anything other than {atomic, _} or {aborted, _} and with the monitor, the explicit catch to stop the connection can never occur so it is probably dead code.

Perhaps it would be nice to have an option for controlling the BEGIN/COMMIT/ROLLBACK timeout, so that this transaction call can be cancelled without the process being killed, but that would be a different feature, solving a different problem.

I might be wrong. Feel free to discuss this further!
